### PR TITLE
Plumb AssignViewsByTagAndType to planner and add tests.

### PR DESCRIPTION
**Pull request by mmandlis**
There is clearly more work on this, but I want to get your feedback first, to make sure I'm in the right direction. Thanks!

### CODE REVIEW

**Comment by mmandlis** [comment]
I merged with the latest changes, PTAL

**Comment by shans** [comment]
I think we want to add a new tactic, like ApplyAll, here. Basically, ApplyAll takes an accumulated set of changes and applies them to a single clone of the recipe. We might want something like ApplyEach (better name maybe?) that clones a recipe for each change, or maybe we'll even need something more complicated to deal with this case.

**Comment by mmandlis** [comment]
Added the most straightforward ApplyEach for now, PTAL.

File: runtime/new-recipe.js [hunk 1]
```diff
@@ -548,21 +551,21 @@ class Walker extends Strategizer.Walker {
...
-        if (result)
+        if (result && (result.constructor != Array || result.length > 0))
           updateList.push({continuation: result, context: particle});
       }
     }
     for (var viewConnection of recipe.viewConnections) {
       if (this.onViewConnection) {
         var result = this.onViewConnection(recipe, viewConnection);
-        if (result)
+        if (result && (result.constructor != Array || result.length > 0))
```
Comment (shans) [inline runtime/new-recipe.js]: this doesn't allow arrays to be returned, but don't we need arrays?

File: runtime/new-recipe.js [hunk 2]
```diff
@@ -571,17 +574,33 @@ class Walker extends Strategizer.Walker {
...
+        case Recipe.Walker.ApplyEach:
+          updateList.forEach(({continuation, context}) => {
+            var cloneMap = new Map();
+            var newRecipe = recipe.clone(cloneMap);
+            if (typeof continuation == 'function')
+              continuation = [continuation];
+            continuation.forEach(f => {
+              f(newRecipe, cloneMap.get(context));
+            });
+            newRecipes.push(newRecipe);
```
Comment (shans) [inline runtime/new-recipe.js]: this looks good

File: runtime/planner.js [hunk 3]
```diff
@@ -87,13 +88,19 @@ class AssignViewsByTagAndType extends Strategy {
...
-          return arc.findViews(view.type, view.tags).map(newView =>
-            ((recipe, viewConnection) => viewConnection.view.id = newView.id));
+          }
+          return arc.findViews(view.type || viewConnection.type, view.tags).map(newView =>
+            ((recipe, viewConnection) => {
+              // TODO: verify that same Arc's view is not assigned to different connections' views.
+              if (newView.type == undefined || viewConnection.type == undefined)
+                viewConnection.connectToView(newView);
+              viewConnection.view.id = newView.id;
+            }));
```
Comment (shans) [inline runtime/planner.js]: this will return an array

File: runtime/new-recipe.js (hunk 1 above)
Comment (mmandlis) [inline runtime/new-recipe.js]: it allows non arrays (result.constructor != Array) OR non empty arrays (result.length > 0)

this is not very readable, i'm not sure how to improve it, maybe move this expression to a  helper method:
if (!isEmpty(result)) {
...
}
where
isEmpty(result) {
  return result && (result.constructor != Array || result.length > 0);
}

**Comment by shans** [comment]
Is this ready to merge? If so, I'll try and hook it up to my strategy explorer before I demo it to you :)

**Comment by mmandlis** [comment]
It is ready and I hooked it into the strategy-explorer demo https://github.com/PolymerLabs/arcs/pull/209

**Pull request merged**